src/foundry/common/documents/macro.mjs.d.ts

```diff
@@ -0,0 +1,194 @@
+// FOUNDRY_VERSION: 10.291
+
+import Document, { DocumentMetadata, DocumentModificationOptions } from "../abstract/document.mjs";
+import * as CONST from "../constants.mjs";
+import * as documents from "./module.mjs";
+import * as fields from "../data/fields.mjs";
+
+declare global {
+  type MacroData = BaseMacro.Properties;
+}
+
+/**
+ * The Document definition for a Macro.
+ * Defines the DataSchema and common behaviors for a Macro which are shared between both client and server.
+ */
+// eslint-disable-next-line @typescript-eslint/no-empty-interface
+interface BaseMacro extends BaseMacro.Properties {}
+declare class BaseMacro extends Document<BaseMacro.SchemaField, BaseMacro.Metadata> {
+  /**
+   * @param data    - Initial data from which to construct the Macro
+   * @param context - Construction context options
+   */
+  constructor(data: BaseMacro.ConstructorData, context?: DocumentConstructionContext);
+
+  static override metadata: Readonly<BaseMacro.Metadata>;
+
+  static override defineSchema(): BaseMacro.Schema;
+
+  /**
+   * The default icon used for newly created Macro documents.
+   */
+  static DEFAULT_ICON: "icons/svg/dice-target.svg";
+
+  override testUserPermission(
+    user: documents.BaseUser,
+    permission: keyof typeof CONST.DOCUMENT_OWNERSHIP_LEVELS | CONST.DOCUMENT_OWNERSHIP_LEVELS,
+    {
+      exact
+    }?: {
+      /**
+       * Require the exact permission level requested?
+       * @defaultValue `false`
+       */
+      exact?: boolean;
+    }
+  ): boolean;
+
+  protected override _preCreate(
+    data: BaseMacro.Properties,
+    options: DocumentModificationOptions,
+    user: foundry.documents.BaseUser
+  ): Promise<void>;
+
+  static override migrateData(source: object): object;
+
+  static override shimData(
+    data: object,
+    {
+      embedded
+    }?: {
+      /**
+       * Apply shims to embedded models?
+       * @defaultValue `true`
+       */
+      embedded?: boolean;
+    }
+  ): object;
+}
+export default BaseMacro;
+
+declare namespace BaseMacro {
+  type Metadata = Merge<
+    DocumentMetadata,
+    {
+      name: "Macro";
+      collection: "macros";
+      indexed: true;
+      compendiumIndexFields: ["_id", "name", "img", "sort"];
+      label: "DOCUMENT.Macro";
+      labelPlural: "DOCUMENT.Macros";
+      coreTypes: CONST.MACRO_TYPES[];
+      permissions: { create: "PLAYER" };
+    }
+  >;
+
+  type SchemaField = fields.SchemaField<Schema>;
+  type ConstructorData = UpdateData;
```

Comment:
Will do. 

Do I also need the generic `<TypeName extends TypeNames>` on these?

Reply:
Another question on this one:
The fields are marked as `required: true` in the Schema interface, but only some are actually required by the constructor, while other required fields get populated with their default value.

Is there a way to tell which ones are actually required by the constructor just by looking at the JSDoc / script file?

I have alternatively been able to test construction over in the runtime to determine which are actually required without producing a validation error, but is that effort necessary, or is there an easier way?

Reply:
> Will do.
> 
> Do I also need the generic `<TypeName extends TypeNames>` on these?

I only really added that when we have actually differing data. I don't think it's going to be that important here to warrant adding it.

> Another question on this one: The fields are marked as `required: true` in the Schema interface, but only some are actually required by the constructor, while other required fields get populated with their default value.
> 
> Is there a way to tell which ones are actually required by the constructor just by looking at the JSDoc / script file?
> 
> I have alternatively been able to test construction over in the runtime to determine which are actually required without producing a validation error, but is that effort necessary, or is there an easier way?

It's basically best to think of it like which setting forbids which values. `required` forbids `undefined`. `nullable`  set to `false` forbids `null` values. Though as long as a field still has an initial value defined, both `null` and `undefined` are allowed. On the macro for example, the name is required, because a `StringField`'s default value is an empty string. But the `blank: false` makes that invalid.